### PR TITLE
RESTEasy Reactive - Support text/binary conversion for multipart files

### LINKED ISSUE #27083: RESTEasy Reactive - Server side multipart implementation is not being respected
### Describe the bug

We have a multipart endpoint which accepts two parts handled as Strings such as this:

```
public class MultipartUpload {

    @RestForm
    @PartType(MediaType.TEXT_PLAIN)
    public String textPart;

    @RestForm
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public String anotherTextPart;

}
```

This is after we migrated to RESTEasy Reactive. With the exception of anotherTextPart being of type InputStream instead nothing has really changed since then, so essentially a multipart where we're not interested in working with the parts as files.

What we weren't aware of was that some of our users are sending us parts with a filename provided, and whenever they do these two parts end up as null.
I've tried looking into the implementation and it kind of looks like the MultipartParser will store everything into a file whenever a filename is provided. Haven't really understood the code completely but my guess is that the implementation will then fail to retrieve the data that was read unless the multipart is defined with File types. But I might be wrong here though.

Looking at the [RFC-2183](https://datatracker.ietf.org/doc/html/rfc2183#section-2.3) for the Content-Disposition header, this behavior does not seem in line with the specification:

> The presence of the filename parameter does not force an
>    implementation to write the entity to a separate file. It is
>    perfectly acceptable for implementations to leave the entity as part
>    of the normal mail stream unless the user requests otherwise. As a
>    consequence, the parameter may be used on any MIME entity, even
>    `inline' ones. These will not normally be written to files, but the
>    parameter could be used to provide a filename if the receiving user
>    should choose to write the part to a file.

The users haven't changed anything in their implementation so it seems pretty evident that this has changed between Classic and Reactive.

### Expected behavior

The server side implementation should be respected regardless of what the client sends.
Unless a RESTEasy server multipart has implemented it, don't write parts of a multipart request to file, even if the Content-Disposition includes a filename. 
The request should still work as if the client had not provided a file name for the part in question, and such is the case in Classic.

### Actual behavior

If a client sends a part with filename in the Content-Disposition header, the receiving end implementation writes the data to a file regardless of what the server side multipart implementation looks like. However, if the server side does not implement the part as a file it will not be able to retrieve the data for further use, i.e the part becomes null.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
Assuming we receive the following request:
```
Content-Disposition: form-data; name="textPart"; filename="my_file.txt"
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary
```

We can load the content of the file using the `FileUpload` class:

```java
public class FormData {
    @FormParam("myFile")
    public FileUpload myFile;
}
```

However, according to what the specs state:
```
The presence of the filename parameter does not force an
implementation to write the entity to a separate file. It is
perfectly acceptable for implementations to leave the entity as part
of the normal mail stream unless the user requests otherwise. As a
consequence, the parameter may be used on any MIME entity, even
`inline' ones. These will not normally be written to files, but the
parameter could be used to provide a filename if the receiving user
should choose to write the part to a file.
```

Then, we should not enforce the use of `FileUpload` and hence also support reading the whole content of the file into strings, and other binary formats like byte[] and InputStream.

```java
public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public String myFile;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public byte[] binaryPart;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public InputStream streamPart;
}
```

Fix https://github.com/quarkusio/quarkus/issues/27083

### CODE REVIEW

File: independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java [hunk 1]
```diff
@@ -867,12 +868,20 @@ public Object getFormParameter(String name, boolean single, boolean encoded) {
         }
         if (single) {
             FormData.FormValue val = formData.getFirst(name);
-            if (val == null || val.isFileItem()) {
+            if (val == null) {
                 return null;
             }
-            if (encoded) {
+
+            if (val.isFileItem()) {
```
Comment (geoand) [inline independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: What case does this change cater to?

Comment (Sgitario) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: When using:
```java
public class FormData {
    @FormParam("myFile")
    public String myFile;
}

```

Comment (Sgitario) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: One idea is to force users to use `@PartType(MediaType.APPLICATION_OCTET_STREAM)` to copy the file content to a string. 
However, as far I understood, this is how resteasy classic works and what the spec says :/

Comment (geoand) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: The spec doesn't say anything about `text` files though, does it? 
All I am questioning here is the need to support `String`. I would much rather not support it initially and have users manually convert the string if necessary.
Or we really want `String`, I think it only makes sense to do so for certain media types.

Comment (Sgitario) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: I've just confirmed that Resteasy Classic works like this. 
If users try to load a binary file to String, I would say this is a user issue, but the framework should not stop this use case. 
Having said that and despite what Resteasy Classic does, I agree with you that we should force users to use the media type `@PartType(MediaType.APPLICATION_OCTET_STREAM)` if they want to copy the content of a file to a String. 

Let me know if you want me to do this.

Comment (geoand) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:875]: Sounds good

Comment (Sgitario) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java]: PR updated

**Review by geoand — APPROVED** [review]

**Pull request merged**